pennylane/fourier/coefficients.py

```diff
@@ -0,0 +1,160 @@
+    degree) of the Fourier spectrum.
+
+    In order to mitigate aliasing, there is an option to apply a low-pass filter
+    prior to computing the coefficients. Coefficients up to a specified value
+    are computed, and then frequencies higher than the degree are simply
+    removed. This ensures that the coefficients returned will have the correct
+    values, though they may not be the full set of coefficients. If no threshold
+    value is provided, the threshold will be set to ``2 * degree``.
+
+    Args:
+        f (callable): function that takes an array of `n_inputs` scalar inputs
+        n_inputs (int): number of function inputs
+        degree (int): max frequency of Fourier coeffs to be computed. For degree :math:`d`,
+            the coefficients from frequencies :math:`-d, -d+1,...0,..., d-1, d` will be computed.
+        lowpass_filter (bool): If ``True``, a simple low-pass filter is applied prior to
+            computing the set of coefficients in order to filter out frequencies above the
+            given degree.
+        filter_threshold (None or int): The integer frequency at which to filter. If
+            ``lowpass_filter`` is set to ``True,`` but no value is specified, ``2 * degree`` is used.
+
+    Returns:
+        array[complex]: The Fourier coefficients of the function ``f`` up to the specified degree.
+
+    **Example**
+
+    .. code-block:: python
+
+        from functools import partial
+        import pennylane as qml
+        from pennylane import numpy as anp
+        from pennylane.fourier import fourier_coefficients
+
+        # Expected Fourier series over 2 parameters with frequencies -1, 0, and 1
+        num_inputs = 2
+        degree = 1
+
+        weights = anp.array([0.5, 0.2])
+
+        dev = qml.device('default.qubit', wires=['a'])
+
+        @qml.qnode(dev)
+        def circuit(weights, inpt):
+            qml.RX(inpt[0], wires='a')
+            qml.Rot(0.1, 0.2, 0.3, wires='a')
+
+            qml.RY(inpt[1], wires='a')
+            qml.Rot(-4.1, 3.2, 1.3, wires='a')
+
+            return qml.expval(qml.PauliZ(wires='a'))
+
+        # Use partial function to compute coefficients of the `inpt` variable
+        coeffs = fourier_coefficients(partial(circuit, weights), num_inputs, degree)
+
+        >>> print(coeffs)
+        [[ 0.    +0.j     -0.    +0.j     -0.    +0.j    ]
+         [-0.0014-0.022j  -0.3431-0.0408j -0.1493+0.0374j]
+         [-0.0014+0.022j  -0.1493-0.0374j -0.3431+0.0408j]]
+    """
+    if not lowpass_filter:
+        return _fourier_coefficients_no_filter(f, n_inputs, degree)
+
+    if filter_threshold is None:
+        filter_threshold = 2 * degree
+
+    # Compute the fft of the function at 2x the specified degree
+    unfiltered_coeffs = _fourier_coefficients_no_filter(f, n_inputs, filter_threshold)
+
+    # Shift the frequencies so that the 0s are at the centre
+    shifted_unfiltered_coeffs = np.fft.fftshift(unfiltered_coeffs)
```

Comment:
This is maybe a crazy question, but would it make sense for this function to be differentiable? I say this because now that #1202 is merged, `qml.math.fft` should be available and support autograd/pytorch/tf.

Reply:
Also, does this function work if the provided QNode used a tf/torch interface(? -- worth double checking, although I remember discussions with @mariaschuld regarding getting it working on other interfaces). If it doesn't, using `qml.math` may mitigate this.

Reply:
Good question - I will look into it!

Reply:
I have tried running `coefficients` for QNodes using both `torch` and `tf` interfaces and it seems to work. Is there any further testing that would need to be done?  

That said, I don't see any reason why not to switch to `qml.math` versions of stuff, provided all the functions used here are available.

Reply:
That should be sufficient!!

If it works using torch and tf, no need to make any changes (does it work with JAX?).

The only reason from here to switch to using `qml.math` is if you want to differentiate `fourier_coefficients`. Is that a use-case that could ever be needed?

Reply:
Did a small test with JAX and that works as well.

I don't know off-hand of any case where we'd need to differentiate the coefficients; maybe @mariaschuld has ideas?

Reply:
Something I immediately think about would be to track the influence of variational parameters in a circuit on the coefficients - but if that would work things would be CRAZY! 

Actually something I need to try next week!

Reply:
But for this module, let's sell it for now as non-diffable (like the kernels module - just some nice utils functions), and then see what we can do!

Reply:
> But for this module, let's sell it for now as non-diffable (like the kernels module - just some nice utils functions), and then see what we can do!

I agree, let's merge this as-is for the next release, and save making it differentiable for a new PR. Will be easier to review as well 😆